plugins/ilitek-its/fu-ilitek-its-device.c

```diff
@@ -0,0 +1,1019 @@
+		return FALSE;
+
+	if (!fu_memread_uint64_safe(rbuf->data, rbuf->len, 0, &version, G_BIG_ENDIAN, error))
+		return FALSE;
+
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		fu_device_set_version_bootloader_raw(FU_DEVICE(self), version);
+
+		/* fw update forcely */
+		fu_device_set_version_raw(FU_DEVICE(self), 0);
+	} else {
+		fu_device_set_version_raw(FU_DEVICE(self), version);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ilitek_its_device_ensure_ic_mode(FuIlitekItsDevice *self, GError **error)
+{
+	guint8 ic_mode;
+	g_autoptr(FuStructIlitekItsHidCmd) st_cmd = fu_struct_ilitek_its_hid_cmd_new();
+	g_autoptr(GByteArray) rbuf = g_byte_array_new();
+
+	fu_struct_ilitek_its_hid_cmd_set_write_len(st_cmd, 1);
+	fu_struct_ilitek_its_hid_cmd_set_read_len(st_cmd, 2);
+	fu_struct_ilitek_its_hid_cmd_set_cmd(st_cmd, FU_ILITEK_ITS_CMD_GET_IC_MODE);
+	if (!fu_ilitek_its_device_send_cmd(self, st_cmd, rbuf, error))
+		return FALSE;
+
+	if (!fu_memread_uint8_safe(rbuf->data, rbuf->len, 0, &ic_mode, error))
+		return FALSE;
+	if (ic_mode == FU_ILITEK_ITS_BL_MODE)
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	else
+		fu_device_remove_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ilitek_its_device_ensure_ic_name_old(FuIlitekItsDevice *self, GError **error)
+{
+	g_autofree gchar *ic_name = NULL;
+	g_autoptr(FuStructIlitekItsHidCmd) st_cmd = fu_struct_ilitek_its_hid_cmd_new();
+	g_autoptr(GByteArray) rbuf = g_byte_array_new();
+
+	fu_struct_ilitek_its_hid_cmd_set_write_len(st_cmd, 1);
+	fu_struct_ilitek_its_hid_cmd_set_read_len(st_cmd, 32);
+	fu_struct_ilitek_its_hid_cmd_set_cmd(st_cmd, FU_ILITEK_ITS_CMD_GET_MCU_VERSION);
+	if (!fu_ilitek_its_device_send_cmd(self, st_cmd, rbuf, error))
+		return FALSE;
+
+	ic_name = g_strdup_printf("%04x", fu_struct_ilitek_its_mcu_version_get_ic_name(rbuf));
+	fu_device_set_name(FU_DEVICE(self), ic_name);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ilitek_its_device_ensure_ic_name(FuIlitekItsDevice *self, GError **error)
+{
+	g_autofree gchar *ic_name = NULL;
+	g_autoptr(FuStructIlitekItsHidCmd) st_cmd = fu_struct_ilitek_its_hid_cmd_new();
+	g_autoptr(GByteArray) rbuf = g_byte_array_new();
+
+	/* check new protocol is supported */
+	if ((fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER) &&
+	     self->protocol_ver < 0x010803) ||
+	    (!fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER) &&
+	     self->protocol_ver < 0x060009))
+		return fu_ilitek_its_device_ensure_ic_name_old(self, error);
+
+	fu_struct_ilitek_its_hid_cmd_set_write_len(st_cmd, 1);
+	fu_struct_ilitek_its_hid_cmd_set_read_len(st_cmd, 32);
+	fu_struct_ilitek_its_hid_cmd_set_cmd(st_cmd, FU_ILITEK_ITS_CMD_GET_MCU_INFO);
+	if (!fu_ilitek_its_device_send_cmd(self, st_cmd, rbuf, error))
+		return FALSE;
+
+	ic_name = fu_struct_ilitek_its_mcu_info_get_ic_name(rbuf);
```

Comment:
```suggestion
	ic_name = g_strdup_printf("Touchscreen IC ILI%s", fu_struct_ilitek_its_mcu_info_get_ic_name(rbuf));
```

How about this? Then it shows up as `Touchscreen IC ILI2901` on our system.

Reply:
Need to modify line: 797 too, would compare name with name in firmware file in prepare_firmware().

Reply:
I can fix this; I think storing both makes sense -- and Name is free to be changed by a daemon/quirk so it's a bit fragile to rely on this for the ic_name.

Reply:
I just modify it to "Touchscreen ILIxxxx" and ic_name is stored.